src/rpc/client.rs

```diff
@@ -0,0 +1,706 @@
+                    if index == tasks.len() {
+                        Loop::Break(result)
+                    } else {
+                        Loop::Continue((index, tasks, result))
+                    }
+                }))
+            }
+        })
+        .map(|r| r.unwrap_or_default())
+    }
+
+    fn region_context(
+        inner: Arc<RpcClientInner>,
+        key: &Key,
+    ) -> impl Future<Item = (RegionContext, Arc<KvClient>), Error = Error> {
+        let inner2 = Arc::clone(&inner);
+        inner
+            .locate_key(key)
+            .and_then(move |location| {
+                let peer = location.peer().expect("leader must exist");
+                let store_id = peer.get_store_id();
+                inner.load_store(store_id).map(|store| RegionContext {
+                    region: location.into_inner(),
+                    store,
+                })
+            })
+            .and_then(move |region| inner2.kv_client(region))
+    }
+
+    fn region_context_by_id(
+        inner: Arc<RpcClientInner>,
+        id: RegionId,
+    ) -> impl Future<Item = (RegionContext, Arc<KvClient>), Error = Error> {
+        let inner2 = Arc::clone(&inner);
+        inner
+            .load_region_by_id(id)
+            .and_then(move |region| {
+                let peer = region.peer().expect("leader must exist");
+                let store_id = peer.get_store_id();
+                inner
+                    .load_store(store_id)
+                    .map(|store| RegionContext { region, store })
+            })
+            .and_then(move |region| inner2.kv_client(region))
+    }
+
+    fn raw(
+        inner: Arc<RpcClientInner>,
+        key: &Key,
+        cf: Option<ColumnFamily>,
+    ) -> impl Future<Item = RawContext, Error = Error> {
+        Self::region_context(inner, key)
+            .map(move |(region, client)| RawContext::new(region, client, cf))
+    }
+
+    fn txn(inner: Arc<RpcClientInner>, key: &Key) -> impl Future<Item = TxnContext, Error = Error> {
+        Self::region_context(inner, key).map(move |(region, _client)| TxnContext::new(region))
+    }
+
+    #[inline]
+    fn inner(&self) -> Arc<RpcClientInner> {
+        Arc::clone(&self.inner)
+    }
+
+    pub fn raw_get(
+        &self,
+        key: Key,
+        cf: Option<ColumnFamily>,
+    ) -> impl Future<Item = Value, Error = Error> {
+        Self::raw(self.inner(), &key, cf)
+            .and_then(move |context| context.client().raw_get(context, key))
+            .and_then(move |value| {
+                if value.is_empty() {
+                    Err(Error::NoSuchKey)
+                } else {
+                    Ok(value)
+                }
+            })
+    }
+
+    pub fn raw_batch_get(
+        &self,
+        keys: Vec<Key>,
+        cf: Option<ColumnFamily>,
+    ) -> impl Future<Item = Vec<KvPair>, Error = Error> {
+        let inner = self.inner();
+        self.group_tasks_by_region(keys)
+            .and_then(move |gk| {
```

Comment:
I'm not sure `gk` is enough context to the reader on what this is without checking `group_tasks_by_region`. It'd be nice to name this better.

Reply:
That's a fine name.